internal/upgrader/upgrader_test.go

```diff
@@ -232,6 +232,98 @@ var _ = Describe("Upgrade", func() {
+					Expect(fakeCFClient.UpgradeServiceInstanceCallCount()).Should(Equal(0))
+				})
+
+				// When `checkUpToDate: true` it takes precedence and the actual value of dryRun should be` irrelevant and lead to the exact same results
+				var upgradeErr2 error
+				upgradeLog2 := captureStdout(func() {
+					l := logger.New(100 * time.Millisecond)
+					defer l.Cleanup()
+					upgradeErr2 = upgrader.Upgrade(fakeCFClient, fakeBrokerName, 1, !passedDryRun, true, l)
+					// Under no circumstances we want an actual upgrade to be scheduled
+					Expect(fakeCFClient.UpgradeServiceInstanceCallCount()).Should(Equal(0))
+				})
+
+				if expectedErr == nil {
+					Expect(upgradeErr).NotTo(HaveOccurred())
+					Expect(upgradeErr2).NotTo(HaveOccurred())
+				} else {
+					Expect(upgradeErr).To(Equal(expectedErr))
+					Expect(upgradeErr2).To(Equal(expectedErr))
+				}
+				if expectedLog == "" {
+					Expect(upgradeLog).To(Equal(""))
+					Expect(upgradeLog2).To(Equal(""))
+				} else {
+					Expect(upgradeLog).To(MatchRegexp(expectedLog))
+					Expect(upgradeLog2).To(MatchRegexp(expectedLog))
+				}
+			},
+			Entry("no plans defined",
+				fmt.Errorf("no service plans available for broker: fake-broker-name"), "",
+				[]ccapi.Plan{},
+				[]ccapi.ServiceInstance{},
+			),
+			Entry("no instances defined",
+				nil, "no instances available to upgrade",
+				[]ccapi.Plan{fakePlan},
+				[]ccapi.ServiceInstance{},
+			),
+			Entry("all instances up to date",
+				nil, "no instances available to upgrade",
+				[]ccapi.Plan{fakePlan},
+				[]ccapi.ServiceInstance{{UpgradeAvailable: false}},
+			),
+			Entry("upgradable instances",
+				fmt.Errorf("check up-to-date failed: found 1 instances which are not up-to-date"),
+				`(\s|.)*`, // we don't care about the logs in this test
+				[]ccapi.Plan{fakePlan},
+				[]ccapi.ServiceInstance{{UpgradeAvailable: true}},
+			),
+			Entry("upgradable instances detailed logs",
+				fmt.Errorf("check up-to-date failed: found 1 instances which are not up-to-date"),
+				getExpectedLogForInstance(getFakeInstanceDetailed()),
+				[]ccapi.Plan{{GUID: getFakeInstanceDetailed().PlanGUID, MaintenanceInfoVersion: getFakeInstanceDetailed().PlanMaintenanceInfoVersion}},
+				[]ccapi.ServiceInstance{getFakeInstanceDetailed()},
+			),
+			Entry("upgradeable instances whose creation failed",
+				nil, "no instances available to upgrade",
+				[]ccapi.Plan{{GUID: fakeInstance1.PlanGUID}},
+				[]ccapi.ServiceInstance{{UpgradeAvailable: true, LastOperation: ccapi.LastOperation{Type: "create", State: "failed"}}},
+			),
+			/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+			// WARNING:
+			// We may want to check whether the following scenarios are possible and if they do, we may want to prevent them from happening
+			/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```

Comment:
```suggestion
```
Discussed these scenarios during standup and we are already investigating some possible corner cases.
Deleting this warning.